import sys
import MySQLdb as mdb
from trangchu import TrangChuCongVan
from PyQt6.QtWidgets import QApplication, QWidget, QMessageBox
from PyQt6.uic import loadUi
from dangki import DangkyApp
from quenmatkhau import QuenMatKhauApp

class LoginApp(QWidget):
    def __init__(self):
        super().__init__()
        loadUi("ui/login.ui", self)
        self.db = None
        self.cursor = None
        self.connect_db()
        self.btn_login.clicked.connect(self.handle_login)
        self.btn_forgot.clicked.connect(self.handle_forgot)
        self.btn_register.clicked.connect(self.handle_register)

    def connect_db(self):
        try:
            self.db = mdb.connect(
                host="localhost",
                user="root",
                passwd="",    
                db="qlcv2",
                charset="utf8"
            )
            self.cursor = self.db.cursor()
            print("‚úÖ K·∫øt n·ªëi MySQL th√†nh c√¥ng!")
        except mdb.Error as e:
            QMessageBox.critical(self, "L·ªói", f"Kh√¥ng th·ªÉ k·∫øt n·ªëi DB!\n{e}")

    def handle_login(self):
        if not self.db:
            QMessageBox.critical(self, "L·ªói", "Ch∆∞a k·∫øt n·ªëi database!")
            return 
        username = self.txt_username.text()
        password = self.txt_password.text()
        try:
            query = "SELECT * FROM user WHERE username=%s AND passwordhash=%s"
            self.cursor.execute(query, (username, password))
            user = self.cursor.fetchone()
            if user:
                QMessageBox.information(self, "Th√†nh c√¥ng", "ƒêƒÉng nh·∫≠p th√†nh c√¥ng!")
                self.hide()
                # L·∫•y userid t·ª´ k·∫øt qu·∫£ truy v·∫•n (gi·∫£ s·ª≠ userid l√† c·ªôt ƒë·∫ßu ti√™n)
                role = user[8] # L·∫•y role t·ª´ c·ªôt th·ª© 9
                fullname = user[3] # L·∫•y fullname t·ª´ c·ªôt th·ª© 4
                self.main_window = TrangChuCongVan(login_window=self, current_user=username, current_role=role, current_fullname=fullname)
                self.main_window.show()
            else:
                QMessageBox.warning(self, "L·ªói", "Sai t√†i kho·∫£n ho·∫∑c m·∫≠t kh·∫©u!")
        except mdb.Error as e:
            QMessageBox.critical(self, "L·ªói", f"L·ªói truy v·∫•n DB: {e}")

    def handle_forgot(self):
        self.hide()
        self.forget_windown = QuenMatKhauApp(self)
        self.forget_windown.show()
    
    def handle_register(self):
        self.hide()
        self.register_window = DangkyApp(self)
        self.register_window.show()

    def closeEvent(self, event):
        if self.db:
            self.cursor.close()
            self.db.close()
            print("üî¥ ƒê√£ ƒë√≥ng k·∫øt n·ªëi DB")
        event.accept()

if __name__ == "__main__":
    app = QApplication(sys.argv)
    window = LoginApp()
    window.show()
    sys.exit(app.exec())
